adr/keycloak-adr.md

```diff
@@ -0,0 +1,236 @@
+#### Architectural Decision Record (ADR): Integration of Keycloak with GitLab for Authentication
+
+#### Status
+Proposed
+
+#### Date
+2024-03-05
+
+
+#### Introduction
+This ADR supports the further integration of Keycloak with GitLab to streamline authentication mechanisms, with the goal of enhancing security and user experience. The drive behind this initiative is to refine login processes, bolster security protocols, and ensure compliance with Information Assurance (IA) standards.
+
+**Objective**: Utilize Keycloak as the predominant Identity and Access Management (IAM) solution alongside GitLab, applying OAuth2 for authentication. This plan aims to consolidate user management and authentication, making credential management more straightforward and improving our security landscape.
+
+**Rationale**: The integration of authentication between Keycloak and GitLab is fundamental for facilitating seamless user access, elevating security via centralized oversight, and adhering to compliance standards. This commitment is built on our adherence to open authentication and authorization standards, such as OAuth2 and OpenID Connect (OIDC).
+
+```mermaid
+graph LR
+  A["User"] -->|1. Requests access to GitLab| C["GitLab"]
+  C -->|2. Redirects to Keycloak| B["Keycloak"]
+  B -->|3. Authenticates user| A
+  B -->|4. Sends authentication token to GitLab| C
+  C -->|5. Grants user access| A
+
+  subgraph "Authentication Flow"
+    B["Keycloak (Identity Provider)"]
+    C["GitLab (Service Provider)"]
+    A["User (Requester)"]
+  end
+```
+
+#### Decision Process
+The decision to further integrate Keycloak with GitLab was informed by evaluating several key factors and prerequisites:
+
+```mermaid
+graph TD
+    A[Start] --> B{License Requirements}
+    B -->|Available: GitLab Premium| C[Proceed with OAuth2/OIDC]
+    B -->|Not Available: GitLab Premium| D[Explore Alternatives]
+    C --> E{Is SAML Integration Needed?}
+    E -->|Yes| F[Defer SAML Integration to Future Phases]
+    E -->|No| G[Proceed with OAuth2/OIDC]
+    G --> H{Interest in Group Integration?}
+    H -->|Yes| I[Plan Group Integration for Future]
+    H -->|No| J[Continue Without Group Integration]
+    I --> K[Conclusion: Document in ADR]
+    J --> K[Conclusion: Document in ADR]
+```
+
+- **SAML Integration**: Our initial focus is on OAuth2/OIDC, with SAML integration to be considered in later phases, allowing us to stay adaptive to technology advancements.
+- **License Requirements**: Presumes availability of GitLab Premium, vital for advanced authentication features, ensuring we can fully leverage GitLab's capabilities.
```

Comment:
What does this give us? What are the tradeoffs?  This sentence doesn't really say anything - OmniAuth/Oauth 2 is part of the free tier of GitLab...